sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java

```diff
@@ -61,18 +79,30 @@ public int writeAsText(ByteBuf buffer, @Nonnull Object value) {
     @Override
     protected byte[] encodeAsUTF8Text(@Nonnull Object value) {
         assert value instanceof String : "value must be a string";
-        return ((String) value).getBytes(StandardCharsets.UTF_8);
+        byte[] bytes = ((String) value).getBytes(StandardCharsets.UTF_8);
+        validateMaxLength(bytes.length);
+        return bytes;
     }
 
     @Override
     public Object readBinaryValue(ByteBuf buffer, int valueLength) {
+        validateMaxLength(valueLength);
         byte[] utf8 = new byte[valueLength];
         buffer.readBytes(utf8);
         return new String(utf8, StandardCharsets.UTF_8);
     }
 
     @Override
     Object decodeUTF8Text(byte[] bytes) {
+        validateMaxLength(bytes.length);
         return new String(bytes, StandardCharsets.UTF_8);
     }
+
+    private void validateMaxLength(int len) {
```

Comment:
I'd remove this from the streaming layer. Once we add proper support for length limits on a type level we'll need this logic elsewhere and here it would just be duplicated.